packages/extension-code-block/src/plugin.ts

```diff
@@ -0,0 +1,45 @@
+// Modified from
+// https://github.com/ProseMirror/prosemirror-inputrules/blob/master/src/inputrules.js
+// license: MIT
+
+import { InputRule } from 'prosemirror-inputrules'
+import { Plugin as ProseMirrorPlugin } from 'prosemirror-state'
+import type { Plugin, TextSelection } from 'prosemirror-state'
+import type { EditorView } from 'prosemirror-view'
+
+const MAX_MATCH = 500
+
+function run(view: EditorView, from: number, to: number, text: string, rules: any[], plugin: Plugin) {
+  if (view.composing) return false
+  const state = view.state; const
+    $from = state.doc.resolve(from)
+  if ($from.parent.type.spec.code) return false
+  const textBefore = $from.parent.textBetween(Math.max(0, $from.parentOffset - MAX_MATCH), $from.parentOffset,
+    undefined, '\ufffc') + text
+  for (let i = 0; i < rules.length; i += 1) {
+    const match = rules[i].match.exec(textBefore)
+    const tr = match && rules[i].handler(state, match, from - (match[0].length - text.length), to)
+    if (!tr) continue
+    view.dispatch(tr.setMeta(plugin, {
+      transform: tr, from, to, text,
+    }))
+    return true
+  }
+  return false
+}
+
+export default function (options: any = {}) {
+  const rules: InputRule[] = options.rules || []
+  const plugin: Plugin = new ProseMirrorPlugin({
+    props: {
```

Comment:
This is missing the `compositionend` event in the original [`inputrule`](https://github.com/ProseMirror/prosemirror-inputrules/blob/master/src/inputrules.js).

Reply:
https://github.com/ProseMirror/prosemirror-inputrules/pull/6

Because marijnh said {enter} is not a kind of inputrule, so I made a standalone plugin

Reply:
The origin inputrule plugin is not removed, IMO there is no necessary to add `compositionend` here

Reply:
> I'm not sure whether `the origin plugin state` is needed here

The plugin state is kept there so that the `undoInputRule` function (which is binded to `backspace`) can work using the `setMeta` and `getMeta` fields.

https://prosemirror.net/docs/ref/#state.Transaction.setMeta
https://prosemirror.net/docs/ref/#state.Transaction.getMeta

> Because marijnh said {enter} is not a kind of inputrule, so I made a standalone plugin

I think you could argue enter is a type of inputrule, but where the trigger isn't an explicit key being entered, but an implicit key being pressed; plus, you're still using the `InputRule` abstraction, just handling / running it differently.